projects/sandbox/train/train/data/base.py

```diff
@@ -0,0 +1,516 @@
+
+    def build_fixed_dataset(
+        self,
+        background: torch.Tensor,
+        waveforms: torch.Tensor,
+        parameters: Dict[str, torch.Tensor],
+        batch_size: int,
+    ):
+        """
+        Method to build a deterministic dataset of injections
+        into background. Useful for validation and testing where
+        we want deterministic results.
+        """
+        # TODO: generalize to multiple background files
+
+        _, rank = self.get_world_size_and_rank()
+
+        # size of psd data + filter settle in + kernel
+        sample_size = int(self.hparams.sample_rate * self.sample_length)
+
+        # size of filter settle in + kernel
+        kernel_size = int(
+            (self.hparams.kernel_length + self.hparams.fduration)
+            * self.hparams.sample_rate
+        )
+
+        # slice waveforms
+        waveforms = self.slice_waveforms(waveforms)
+
+        # calculated number of background kernels we can make
+        # depending on the stride size and length of validation set
+        stride_size = int(self.hparams.val_stride * self.hparams.sample_rate)
+        num_kernels = (background.shape[-1] - kernel_size) // stride_size + 1
+        num_kernels = int(num_kernels)
+
+        # Create pre-computed kernels of pure background
+        # slice our background so that it has an integer number of
+        # windows, then add dummy dimensions since unfolding only
+        # works on 4D tensors.
+        # Add a shift corresponding to current rank (if distributed)
+        # to add more diversity of background data in validation
+        shift = int(self.hparams.sample_rate * rank)
+
+        background = background[
+            :, : int(num_kernels * stride_size + sample_size + shift)
+        ]
+        background = background[:, shift:]
+        background = unfold_windows(
+            background, sample_size, stride=stride_size
+        )
+
+        # create repeats of background kernels if we need any
+        repeats = math.ceil(len(waveforms) / len(background))
+        background = background.repeat(repeats, 1, 1)
+        background = background[: len(waveforms)]
+
+        # calculate psd, inject and whiten
+        X, psds = self.psd_estimator(background)
+        X += waveforms
+        X = self.whitener(X, psds)
+
+        # downselect to requested inference parameters and
+        # stack into a single tensor, and scale
+        parameters = {
+            k: v for k, v in parameters.items() if k in self.inference_params
+        }
+        parameters = [torch.Tensor(v) for v in parameters.values()]
+        parameters = torch.vstack(parameters).T
+        dataset = torch.utils.data.TensorDataset(X, parameters)
+
+        return torch.utils.data.DataLoader(
+            dataset,
+            pin_memory=False,
+            batch_size=batch_size,
+            num_workers=self.num_workers,
+        )
+
+    def train_dataloader(self):
+        dataset = Hdf5TimeSeriesDataset(
```

Comment:
I'm seeing the source code of the Hdf5 dataset. This does not preserve existing behavior of pushing the entire background to the GPU. From what I understand this will load background on the CPU and do a transfer to the GPU. Is that correct?